docs/fp-core.md

```diff
@@ -0,0 +1,90 @@
+# Finality Provider Core
+
+This doc is to provide basic invariants and heuristics one should follow to
+build the finality provider. This part should be agnostic to any consumer
+chains.
+
+## Status Transitions
+
+The lifecycle of a finality provider instance involves 5 status: `Registered`,
+`Inactive`, `Active`, `Jailed`, `Slashed`. The status transition is depicted
+in the following diagram.
+
+![Finality Provider Status Transition](./fp-status-transition.png).
+
+## Invariants
+
+We define invariants that the finality provider should satisfy:
+
+- Finality signature should be submitted for a block if:
+  - the block is not finalized,
+  - the finality provider has positive voting power for this block,
+  - the finality provider has not cast a finality signature for this block, and
+  - the finality provider has cast finality votes for all non-finalized blocks
+    prior to this block for which it has voting power.
+
+- Public randomness should be committed for a block if:
+  - the block is not finalized, and
```

Comment:
Is this constraint necessary and did we implement it?

Reply:
This indicates that if the block is already finalized, we can skip it but it does not mean that we cannot commit pub rand for a finalized block. We have implemented in https://github.com/babylonlabs-io/finality-provider/blob/8c63dae68284cb5426cb9cc07967e367c7d627cd/finality-provider/service/fp_instance.go#L502